File: python/y2024/day04.py
from aoc.core import solve, AocInput
from collections import defaultdict
from dataclasses import astuple, dataclass
from functools import partial


@dataclass(frozen=True, slots=True)
class Point:
    x: int
    y: int

    def __iter__(self):
        return iter(astuple(self))

    @staticmethod
    def North():
        return Point(0, -1)

    @staticmethod
    def NorthEast():
        return Point(1, -1)

    @staticmethod
    def East():
        return Point(1, 0)

    @staticmethod
    def SouthEast():
        return Point(1, 1)

    @staticmethod
    def South():
        return Point(0, 1)

    @staticmethod
    def SouthWest():
        return Point(-1, 1)

    @staticmethod
    def West():
        return Point(-1, 0)

    @staticmethod
    def NorthWest():
        return Point(-1, -1)

    def __add__(self, point: "Point") -> "Point":
        if isinstance(point, Point) is False:
            return NotImplemented
        return Point(self.x + point.x, self.y + point.y)

    def __mul__(self, factor: int) -> "Point":
        if isinstance(factor, int) is False:
            return NotImplemented
        return Point(self.x * factor, self.y * factor)


Map = defaultdict[Point, str]


def parse_map(lines: list[str]) -> Map:
    return defaultdict(str, {Point(x, y): c for y, l in enumerate(lines) for x, c in enumerate(l)})


def search_word(map: Map, pattern: str, p: Point, dir: Point) -> bool:
    chars = [map.get(p + dir * i) for i in range(0, len(pattern))]
    return chars == list(pattern) or chars == list(reversed(pattern))


def search_xmas(map: Map) -> list[bool]:
    directions = [Point.East(), Point.SouthEast(), Point.South(), Point.SouthWest()]
    search_map_for_xmas = partial(search_word, map, "XMAS")
    return [search_map_for_xmas(p, dir) for p in map.keys() for dir in directions]


def search_xmas_shape(map: Map) -> list[bool]:
    search_map_for_xmas_shape = partial(search_word, map, "MAS")
    return [
        search_map_for_xmas_shape(p + Point.NorthWest(), Point.SouthEast())
        and search_map_for_xmas_shape(p + Point.SouthWest(), Point.NorthEast())
        for p in map.keys()
    ]


def part1(input: AocInput) -> int:
    return sum(search_xmas(parse_map(input.lines)))


def part2(input: AocInput) -> int:
    return sum(search_xmas_shape(parse_map(input.lines)))


def main() -> None:
    YEAR = "y2024"
    DAY = "Day04"
    print(f"Running year {YEAR.replace('y', '')} day {DAY.replace('Day', '')}")
    solve(YEAR, DAY, part1, 2358, True)
    solve(YEAR, DAY, part2, 1737, True)

File: dotnet/2024/AdventOfCode.Y2024/Day04.cs
using System.Collections.Frozen;
using AdventOfCode.Lib;
using Map = System.Collections.Frozen.FrozenDictionary<AdventOfCode.Lib.Point, char>;

namespace AdventOfCode.Y2024;

[AocPuzzle(2024, 4, "Ceres Search", "C#")]
public sealed class Day04 : IAocDay<int>
{
    public static int Part1(AocInput input) =>
        input.AllLines
            .Pipe(ParseMap)
            .Pipe(
                map =>
                    from point in map.Keys
                    from direction in new[] { Point.East, Point.SouthEast, Point.South, Point.SouthWest }
                    select SearchWord(map, "XMAS", point, direction)
            )
            .Pipe(matches => matches.Count(m => m));

    public static int Part2(AocInput input) =>
        input.AllLines
            .Pipe(ParseMap)
            .Pipe(
                map =>
                    from point in map.Keys
                    select SearchWord(map, "MAS", point + Point.NorthWest, Point.SouthEast) &&
                           SearchWord(map, "MAS", point + Point.SouthWest, Point.NorthEast)
            )
            .Pipe(matches => matches.Count(m => m));

    private static bool SearchWord(Map map, string pattern, Point point, Point direction)
    {
        var len = pattern.Length;
        char[] chars = [..Enumerable.Range(0, len).Select(i => map.GetValueOrDefault(point + direction * i))];
        return chars.SequenceEqual(pattern) || chars.SequenceEqual(pattern.Reverse());
    }

    private static Map ParseMap(string[] lines) => (
        from y in Enumerable.Range(0, lines.Length)
        from x in Enumerable.Range(0, lines[0].Length)
        select KeyValuePair.Create(new Point(x, y), lines[y][x])
    ).ToFrozenDictionary();
}